Composer/packages/tools/language-servers/language-generation/src/LGServer.ts

```diff
@@ -14,37 +14,44 @@ import {
   CompletionItemKind,
   CompletionItem,
   Range,
+  DiagnosticSeverity,
 } from 'vscode-languageserver-types';
 import { TextDocumentPositionParams } from 'vscode-languageserver-protocol';
 import get from 'lodash/get';
-import { LGTemplate, Diagnostic as LGDiagnostic } from 'botbuilder-lg';
+import { LgFile, LgTemplate, lgIndexer, filterTemplateDiagnostics, isValid } from '@bfc/indexers';
 
 import { buildInfunctionsMap } from './builtinFunctionsMap';
 import {
   getRangeAtPosition,
-  getLGResources,
   updateTemplateInContent,
-  getTemplateRange,
   LGDocument,
-  checkText,
   checkTemplate,
   convertDiagnostics,
-  isValid,
-  TRange,
+  generageDiagnostic,
+  LGOption,
+  parse,
+  LGCursorState,
 } from './utils';
 
+const { check } = lgIndexer;
+
 // define init methods call from client
 const InitializeDocumentsMethodName = 'initializeDocuments';
 
-const allowedCompletionStates = ['expression'];
+const { ROOT, TEMPLATENAME, TEMPLATEBODY, EXPRESSION, COMMENTS, SINGLE, DOUBLE } = LGCursorState;
 
 export class LGServer {
   protected workspaceRoot?: URI;
   protected readonly documents = new TextDocuments();
   protected readonly pendingValidationRequests = new Map<string, number>();
-  protected LGDocuments: LGDocument[] = []; // LG Documents Store
-
-  constructor(protected readonly connection: IConnection) {
+  /**
+   * LG documents infomation store.
+   * The connection between lgOption and botProjectService.
+   * Help do inline template editing and server resource passing.
+   */
+  protected LGDocuments: LGDocument[] = [];
+
+  constructor(protected readonly connection: IConnection, protected readonly botProjectService?) {
```

Comment:
I understand the intention is to leverage the API server to get LG file content, instead of passing the whole content directly through LSP protocol. 

Although this is a specific change for Composer, we should seek for a generic solution based on a generic interface, to make sure the modules are better decoupled.

In this case, the suggestion is to let LGServer accept a generic delegate\callback\interface which can get a LGFile given a file id, then implement this delegate using BotProjectService instead of passing it in. 

For example, let's say "FileResolver(id) => string or LGFile". Creating such an interface will force you to think, what's the precise and minimal requirement from LSP to the external world.

Reply:
ok, it's a good idea, i will rewrite this part.